### ISSUE TITLE
quippy-ase conversion problems in gradients

### ISSUE THREAD

**Issue by imagdau** [libAtoms/QUIP#453]
Gradient indices ['ii'] and gradient mask ['has_data'] returned in quippy, are incorrect.
The minimum working example attached compares quippy output to quip output.
This was tested on both Linux and MacOS.
[Archive.zip](https://github.com/libAtoms/QUIP/files/9249900/Archive.zip)

**Comment by albapa** [comment]
Hi Ioan,

quippy uses a different convention (when returning the gradient data) to the Fortran descriptor container. The key fields are `grad_data`, `grad_index_0based` and, if you need it, `pos`.

In your example:
- `data["grad_data"].shape` is `(2655, 3, 1501)` which means that in total there are 2655 derivatives of descriptors w.r.t. neighbours, in 3 Cartesian dimensions, and the descriptor size is 1501.
- `data["grad_index_0based"].shape` is `(2655, 2)` and contains information of all of the gradients in the rows: column 0 is the descriptor index (in case of SOAP, this coincides with the atom index, but it doesn't have to), column 1 is the index of the neighbouring atom the descriptor was differentiated with respect to.
- `data["pos"].shape` is `(2655, 3)`, again, mapping on all gradients and containing the difference vectors, which is useful if you want to assemble virials.

I'd like to understand why `"ii"` is corrupted, but I think you should, for the time being, regard it as superfluous, since you have all the info you need. Hope this helps.

**Comment by albapa** [comment]
Also note that the header of the `quip` output is not consistent with the lines below - sorry for that: the `derivative w.r.t. Cartesian dimension` column is missing, although it is n the right order. I'll fix that as well.

**Comment by albapa** [comment]
@jameskermode having said that, it is unclear why and how the `"ii"` field gets corrupted - since the `"grad_index_0based"` field is filled from the very same `"ii"` field in `quippy/quippy/descriptors.py`.

**Comment by albapa** [comment]
Unless `descriptor_data_mono_to_dict` only copies the pointer address and somehow the target (probably at the Fortran level?) gets deallocated. The other fields are all passed through some `np.` operation which means that they get their own storage freshly allocated and data is explicitly copied.

@imagdau let me know what you need exactly. If you don't need `"ii"` or the other stuff, then we should remove the keys in `descriptors.py`, otherwise I am happy to keep them in some form, but we need to make sure they stay.

**Comment by jameskermode** [comment]
> Unless descriptor_data_mono_to_dict only copies the pointer address and somehow the target (probably at the Fortran level?) gets deallocated. The other fields are all passed through some np. operation which means that they get their own storage freshly allocated and data is explicitly copied.

Yes, this makes sense - Fortran arrays are exposed to Python directly, not copied.

**Comment by jameskermode** [comment]
@stenczelt and @imagdau can this be closed now #553 is merged?

**Comment by Sideboard** [comment]
With the current public branch (QUIP bae81c5f2, GAP b6756dd,  soap_turbo 6a94c14) `test_grad_ii` fails on my laptop. The `ii` array is non-sensical, probably unallocated.
```
ii [array([-406584304,      21910], dtype=int32), array([-406578673,      21910], dtype=int32), array([-1100963216,       21907, -1090411040], dtype=int32)]
grad_data [[0 0]
 [0 2]
 [1 1]
 [1 2]
 [2 2]
 [2 1]
 [2 0]]

FAIL: test_grad_ii (test_descriptor.Test_Descriptor)
Traceback (most recent call last):
  File "$QUIP_ROOT/tests/test_descriptor.py", line 244, in test_grad_ii
    assert ii_item in grad_data
AssertionError
```

```
$ uname -a
Linux cos8-a54ea445 4.18.0-489.el8.x86_64 #1 SMP Thu Apr 27 17:02:11 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
$ gfortran --version
GNU Fortran (GCC) 8.5.0 20210514 (Red Hat 8.5.0-19)
```
Edit: If I copy & paste the test setup in an interactive python shell, it works fine.
```
>>> import quippy
>>> import ase
>>> at_C2H = ase.Atoms(positions=[[0., 0., 0.], [0.875, 0.875, 0.875], [0.2, 0.2, 0.1]],
...                                 symbols=['C', 'C', 'H'],
...                                 cell=[[0.0, 1.75, 1.75], [1.75, 0.0, 1.75], [1.75, 1.75, 0.0]],
...                                 pbc=True)
>>> desc = quippy.descriptors.Descriptor(
...             "soap cutoff=1.3 l_max=4 n_max=4 atom_sigma=0.5 n_Z=2 Z={1 6}"
...         )
>>> data = desc.calc(at_C2H, grad=True)
>>> data['ii']
[array([1, 3], dtype=int32), array([2, 3], dtype=int32), array([3, 2, 1], dtype=int32)]
```

**Comment by albapa** [comment]
I think the problem is that we don't copy the fortran array and it gets deallocated at some point.

**Comment by jameskermode** [comment]
but wasn't that fixed in #553 ?
